misc/toml-parser/src/main/java/sementic/TomlTransformer.java

```diff
@@ -0,0 +1,382 @@
+
+import io.ballerina.tools.text.LinePosition;
+import io.ballerina.tools.text.LineRange;
+import sementic.nodes.DiagnosticLog;
+import sementic.nodes.DiagnosticPos;
+import sementic.nodes.DiagnosticSource;
+import sementic.nodes.TomlArray;
+import sementic.nodes.TomlBasicValue;
+import sementic.nodes.TomlKey;
+import sementic.nodes.TomlKeyValue;
+import sementic.nodes.TomlNode;
+import sementic.nodes.TomlTable;
+import sementic.nodes.TomlTableArray;
+import sementic.nodes.TomlValue;
+import sementic.nodes.TopLevelNode;
+import sementic.tools.DiagnosticCode;
+import syntax.tree.Array;
+import syntax.tree.BasicValueNode;
+import syntax.tree.KeyValue;
+import syntax.tree.ModuleMemberDeclarationNode;
+import syntax.tree.ModulePartNode;
+import syntax.tree.Node;
+import syntax.tree.NodeList;
+import syntax.tree.NodeTransformer;
+import syntax.tree.SeparatedNodeList;
+import syntax.tree.SyntaxKind;
+import syntax.tree.TableArrayNode;
+import syntax.tree.TableNode;
+import syntax.tree.Token;
+import syntax.tree.ValueNode;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Transformer to transform Syntax tree into Abstract Syntax Tree.
+ */
+public class TomlTransformer extends NodeTransformer<TomlNode> {
+
+    private DiagnosticSource diagnosticSource;
+    private DiagnosticLog dlog;
+
+    public TomlTransformer(DiagnosticSource diagnosticSource) {
+
+        this.diagnosticSource = diagnosticSource;
+        this.dlog = DiagnosticLog.getInstance();
+    }
+
+    @Override
+    public TomlNode transform(ModulePartNode modulePartNode) {
+        TomlKey tomlKey = new TomlKey("RootNode", SyntaxKind.MODULE_NAME);
+        TomlTable rootTable = new TomlTable(tomlKey);
+
+        NodeList<ModuleMemberDeclarationNode> members = modulePartNode.members();
+        for (ModuleMemberDeclarationNode rootNode : members) {
+            TomlNode transformedChild = rootNode.apply(this); //todo dotted
+            addChildNodeToParent(rootTable, transformedChild);
+        }
+        return rootTable;
+    }
+
+    private void addChildNodeToParent(TomlTable rootTable, TomlNode transformedChild) {
+        if (transformedChild instanceof TopLevelNode) {
+            if (transformedChild instanceof TomlTable) {
+                TomlTable tableChild = (TomlTable) transformedChild;
+                addChildTableToParent(rootTable, tableChild);
+            } else if (transformedChild instanceof TomlTableArray) {
+                TomlTableArray transformedArray = (TomlTableArray) transformedChild;
+                addChildParentArrayToParent(rootTable, transformedArray);
+            } else if (transformedChild instanceof TomlKeyValue) {
+                TomlKeyValue transformedKeyValuePair = (TomlKeyValue) transformedChild;
+                addChildKeyValueToParent(rootTable, transformedKeyValuePair);
+            } else {
+                throw new UnsupportedOperationException();
+            }
```

Comment:
Can replace the `instanceof` checks with the syntax kind checks. Then can use a switch-case as well.